### PR TITLE
7591_music_offline

### PR DESCRIPTION
**Pull request by vietdung1996**
## [Pull request title]
7591_music_offline
### [Redmine link]
https://edu-redmine.sun-asterisk.vn/issues/7591

### CODE REVIEW

**Comment by vietdung1996** [comment]
ready

File: app/src/main/java/com/example/dung/music_59/data/source/local/LoadTrackLocalAsync.java [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+        mContext = context;
+        mCallBack = callBack;
+    }
+
+    @Override
+    protected List<Track> doInBackground(String... strings) {
+        return loadTrackData();
+    }
+
+    private List<Track> loadTrackData() {
```
Comment (NguyenMinhTuanB) [inline app/src/main/java/com/example/dung/music_59/data/source/local/LoadTrackLocalAsync.java]:
```suggestion
    private List<Track> loadLocalTracks() {
```

File: app/src/main/java/com/example/dung/music_59/data/source/local/LoadTrackLocalAsync.java [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+        ContentResolver cr = mContext.getContentResolver();
```
Comment (NguyenMinhTuanB) [inline app/src/main/java/com/example/dung/music_59/data/source/local/LoadTrackLocalAsync.java]: Phần xử lý đọc nhạc bằng ContentProvider nên viết rõ ràng, tường minh.
Example:
private static final String FILE_FORMAT = "!=0";
    public List<Song> getAllMySongs() {
        Uri songUri = MediaStore.Audio.Media.EXTERNAL_CONTENT_URI;
        String[] projections = new String[]{
                MediaStore.Audio.AudioColumns.TITLE,
                MediaStore.Audio.AudioColumns.DATA,
                MediaStore.Audio.AudioColumns.ARTIST,
                MediaStore.Audio.AudioColumns.DURATION,
                MediaStore.Audio.AudioColumns.ALBUM
        };
        String selection = MediaStore.Audio.Media.IS_MUSIC + FILE_FORMAT;
        Cursor c = mContext.getContentResolver().query(songUri,
                projections, selection, null, null);
        return getAllSongFromCursor(c);
    }

    private List<Song> getAllSongFromCursor(Cursor c) {
        List<Song> songs = new ArrayList<>();
        if (c == null) {
            return songs;
        }
        c.moveToFirst();
        int titleIndex = c.getColumnIndex(MediaStore.Audio.AudioColumns.TITLE);
        int dataIndex = c.getColumnIndex(MediaStore.Audio.AudioColumns.DATA);
        int artistIndex = c.getColumnIndex(MediaStore.Audio.AudioColumns.ARTIST);
        int durationIndex = c.getColumnIndex(MediaStore.Audio.AudioColumns.DURATION);
        int albumIndex = c.getColumnIndex(MediaStore.Audio.AudioColumns.ALBUM);
        while (!c.isAfterLast()) {
            Song song = new Song();
            String title = c.getString(titleIndex);
            String uri = c.getString(dataIndex);
            String artist = c.getString(artistIndex);
            int duration = c.getInt(durationIndex);
            String album = c.getString(albumIndex);

            song.setTitle(title);
            song.setUri(uri);
            song.setUsername(artist);
            song.setDuration(duration);
            song.setAlbum(album);
            songs.add(song);
            c.moveToNext();
        }
        c.close();
        return songs;

**Comment by vietdung1996** [comment]
ready

File: app/src/main/java/com/example/dung/music_59/data/source/TrackDataSource.java [hunk 3]
```diff
@@ -12,11 +12,19 @@
 
     interface OnGetTrackCallBack {
         void onTrackLoaded(List<Track> tracks);
```
Comment (NguyenMinhTuanB) [inline app/src/main/java/com/example/dung/music_59/data/source/TrackDataSource.java]:
```suggestion
        void onTracksLoaded(List<Track> tracks);
```

File: app/src/main/java/com/example/dung/music_59/data/source/TrackDataSource.java [hunk 4]
```diff
@@ -12,11 +12,19 @@
...
+
         void onFailure();
```
Comment (NguyenMinhTuanB) [inline app/src/main/java/com/example/dung/music_59/data/source/TrackDataSource.java]: Lỗi thì nên trả về một Exception

File: app/src/main/java/com/example/dung/music_59/data/source/TrackDataSource.java [hunk 5]
```diff
@@ -12,11 +12,19 @@
...
     }
```
Comment (NguyenMinhTuanB) [inline app/src/main/java/com/example/dung/music_59/data/source/TrackDataSource.java:18]: remove blank line

File: app/src/main/java/com/example/dung/music_59/data/source/local/LoadTrackLocalAsync.java [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+        mContext = context;
+        mCallBack = callBack;
+    }
+
+    @Override
+    protected List<Track> doInBackground(String... strings) {
+        return getAllMySongs();
+    }
+
+    public List getAllMySongs() {
```
Comment (NguyenMinhTuanB) [inline app/src/main/java/com/example/dung/music_59/data/source/local/LoadTrackLocalAsync.java]: copy code rồi. code anh đưa cho em dưới Object là Song
Còn em đang dùng là Track

**Comment by vietdung1996** [comment]
ready

File: app/src/main/java/com/example/dung/music_59/data/source/local/TrackLocalDataSource.java [hunk 7]
```diff
@@ -23,4 +23,9 @@ public static TrackLocalDataSource getInstance(Context context) {
     public void getGenres(TrackDataSource.OnGetGenresCallBack callBack) {
         new GenresAsyncTask(mContext, callBack).execute();
     }
+
+    @Override
+    public void getTrackLocal(TrackDataSource.OnGetTrackLocalCallBack callBack) {
+        new LoadTrackLocalAsync(mContext,callBack).execute();
```
Comment (NguyenMinhTuanB) [inline app/src/main/java/com/example/dung/music_59/data/source/local/TrackLocalDataSource.java]: format this code and all class

**Comment by vietdung1996** [comment]
ready

**Pull request merged**